src/lib/operators/abstract_operator.cpp

```diff
@@ -122,19 +161,47 @@ void AbstractOperator::execute() {
 
 std::shared_ptr<const Table> AbstractOperator::get_output() const { return _output; }
 
-void AbstractOperator::clear_output() { _output = nullptr; }
+void AbstractOperator::clear_output() {
+  Assert(_consumer_count == 0, "Cannot clear output since there are still consuming operators.");
+  if (_clear_output) _output = nullptr;
+}
 
 std::string AbstractOperator::description(DescriptionMode description_mode) const { return name(); }
 
-std::shared_ptr<AbstractOperator> AbstractOperator::deep_copy() const {
-  std::unordered_map<const AbstractOperator*, std::shared_ptr<AbstractOperator>> copied_ops;
-  return _deep_copy_impl(copied_ops);
+std::shared_ptr<AbstractOperator> AbstractOperator::deep_copy(
+    std::unordered_map<const AbstractOperator*, std::shared_ptr<AbstractOperator>>& copied_ops) const {
+  const auto copied_ops_iter = copied_ops.find(this);
+  if (copied_ops_iter != copied_ops.end()) return copied_ops_iter->second;
+
+  const auto copied_left_input =
+      left_input() ? left_input()->deep_copy(copied_ops) : std::shared_ptr<AbstractOperator>{};
+  const auto copied_right_input =
+      right_input() ? right_input()->deep_copy(copied_ops) : std::shared_ptr<AbstractOperator>{};
+
+  auto copied_op = _on_deep_copy(copied_left_input, copied_right_input, copied_ops);
+  if (_transaction_context) copied_op->set_transaction_context(*_transaction_context);
+
+  copied_ops.emplace(this, copied_op);
```

Comment:
Does not work because we need to return `copied_op` as well.

Reply:
Just adding `std::move` results in null pointer exceptions. Nevertheless, it works as follows:
```cpp
std::shared_ptr<AbstractOperator> AbstractOperator::deep_copy(
    std::unordered_map<const AbstractOperator*, std::shared_ptr<AbstractOperator>>& copied_ops) const {
  
  [...]

  copied_ops.emplace(this, std::move(copied_op));

  return copied_ops.at(this);
}
```

But, since we talk about shared pointer copies only, where do you see an advantage?

Reply:
Ah, crap. Didn't realized that we returned `copied_op`, too. Please revert it to the original version (no move). Mea culpa.